pkg/visor/rpc_client.go

```diff
@@ -223,6 +233,100 @@ func (rc *rpcClient) Loops() ([]LoopInfo, error) {
 	return loops, err
 }
 
+// RPCClientDialer keeps track of an rpc connection and retries to connect if it fails at some point
+type RPCClientDialer struct {
+	dialer *snet.Network
+	pk     cipher.PubKey
+	port   uint16
+	conn   net.Conn
+	mu     sync.Mutex
+	done   chan struct{} // nil: loop is not running, non-nil: loop is running.
+}
+
+// NewRPCClientDialer creates a new RPCDialer to the given address
+func NewRPCClientDialer(dialer *snet.Network, pk cipher.PubKey, port uint16) *RPCClientDialer {
+	return &RPCClientDialer{
+		dialer: dialer,
+		pk:     pk,
+		port:   port,
+	}
+}
+
+// Run repeatedly dials to remote until a successful connection is established.
+// It exposes a RPC Server.
+// It will return if Close is called or crypto fails.
+func (d *RPCClientDialer) Run(srv *rpc.Server, retry time.Duration) error {
+	if ok := d.setDone(); !ok {
+		return ErrAlreadyServing
+	}
+	for {
+		if err := d.establishConn(); err != nil {
+			return err
+		} else {
```

Comment:
I would get rid of an `else` statement here because there's a `return` statement in the `if` block above.